godot-codegen/src/models/domain/enums.rs

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Definition of engine enums/bitfields and related types.
+//!
+//! See also generator/enums.rs for functions related to turning enums into `TokenStream`s.
+
+use crate::util::ident;
+
+use proc_macro2::{Ident, Literal, TokenStream};
+use quote::{quote, ToTokens};
+
+pub struct Enum {
+    pub name: Ident,
+    pub godot_name: String,
+    pub is_bitfield: bool,
+    pub enumerators: Vec<Enumerator>,
+}
+
+impl Enum {
+    /// Which derives should be implemented for this enum.
+    pub fn derives(&self) -> Vec<Ident> {
+        let mut derives = vec!["Copy", "Clone", "Eq", "PartialEq", "Hash", "Debug"];
+
+        if self.is_bitfield {
+            derives.push("Default");
+        }
+
+        derives.into_iter().map(ident).collect()
+    }
+
+    /// The type we use to represent values of this enum.
+    pub fn ord_type(&self) -> Ident {
+        if self.is_bitfield {
+            ident("u64")
+        } else {
+            ident("i32")
+        }
+    }
+
+    /// Returns all unique enumerator ords, sorted.
+    ///
+    /// Returns `None` if `self` is a bitfield.
+    pub fn unique_ords(&self) -> Option<Vec<i32>> {
+        let mut unique_ords = self
+            .enumerators
+            .iter()
+            .map(|enumerator| enumerator.value.as_enum())
+            .collect::<Option<Vec<i32>>>()?;
+
+        unique_ords.sort();
+        unique_ords.dedup();
+
+        Some(unique_ords)
+    }
+
+    /// Returns tokens representing the engine trait this enum should implement.
+    pub fn engine_trait(&self) -> TokenStream {
+        if self.is_bitfield {
+            quote! { crate::obj::EngineBitfield }
+        } else {
+            quote! { crate::obj::EngineEnum }
+        }
+    }
+
+    /// Returns the maximum index of an indexable enum.
+    ///
+    /// Return `None` if `self` isn't an indexable enum. Meaning it is either a bitfield, or it is an enum that can't be used as an index.
+    pub fn count_index_enum(&self) -> Option<usize> {
```

Comment:
The `try_` expressed that it tried to count, and if it's not an index enum, it would return `None`.

But it's not the greatest name... maybe something entirely different like `find_index_enum_max` is appropriate? Where "find" implies the optionality.